addons/payment_toss_payments/controllers/main.py

```diff
@@ -0,0 +1,127 @@
+        return request.redirect('/payment/status')
+
+    @http.route(_fail_url, type='http', auth='public')
+    def handleValidationFail(self, **error_data):
+        """ Handle unsuccessful validation from Toss SDK.
+
+        :param dict error_data: The dict data of successUrl's query parameter.
+        Expected keys = ["code", "message", "orderId"]
+        """
+        tx_sudo = request.env['payment.transaction'].sudo() \
+            ._search_by_reference('toss_payments', error_data)
+
+        if not tx_sudo:
+            _logger.error("Cannot find matching payment trasaction using reference: %s",
+                          error_data.get('orderId'))
+            return request.redirect('/payment/status')
+
+        _logger.error(
+            "Unsuccessful payment validation. Error code: %s, Error message: %s, Order ID: %s",
+            error_data.get('code'),
+            error_data.get('message'),
+            error_data.get('orderId'),
+        )
+        tx_sudo._set_error(error_data.get('message'))
+
+        return request.redirect('/payment/status')
+
+    @http.route(_webhook_url, type='http', auth='public', methods=['POST'], csrf=False)
+    def toss_payments_webhook(self):
+        """ Process the notification data sent by Toss Payments to the webhook.
+
+        Event message schema can be found in Toss Payments documentation:
+        https://docs.tosspayments.com/reference/using-api/webhook-events#%EC%9D%B4%EB%B2%A4%ED%8A%B8-%EB%B3%B8%EB%AC%B8
+
+        :param dict data: The event data.
+        :return: The 'OK' string to acknowledge the notification.
+        :rtype: str
+        """
+        request_data = request.get_json_data()
+        eventType = request_data.get('eventType')
+
+        if eventType == 'PAYMENT_STATUS_CHANGED':
+            payment_data = request_data.get('data')
+            tx_sudo = request.env['payment.transaction'].sudo() \
+                ._search_by_reference('toss_payments', payment_data)
+            self._verify_signature(tx_sudo, payment_data)
+            tx_sudo._apply_updates(payment_data)
+            return 'OK'
+
+        _logger.warning("Webhook endpoint received unknown event %s", eventType)
+        msg = "Unrecognized event type"
+        raise BadRequest(msg)
+
+    @staticmethod
+    def _verify_signature(tx_sudo, payment_data):
+        """ Check that the received payment data's secret key matches the tx_sudo's secret key.
+
+        :param payment.transaction tx_sudo: The sudoed transaction referenced by the payment data.
+        :param dict payment_data: The payment data.
+        :return: None
+        :raise Forbidden: If the payment secret keys don't match.
+        """
+        # Expired event might not have secret if we never initiated the payment flow
+        if payment_data.get('status') == 'EXPIRED':
+            return
+
+        if not (payment_secret := payment_data.get('secret')):
+            msg = "Received payment data with missing secret key."
```

Comment:
Consider the same warning [messages](https://github.com/odoo/odoo/blob/efc2788dfccd13ee6feb309430ff57e49664ff97/addons/payment_buckaroo/controllers/main.py#L96) to align with all the other modules.